packages/story-editor/src/components/panels/design/layer/useLayerSelection.js

```diff
@@ -22,75 +22,67 @@ import { useCallback } from '@web-stories-wp/react';
 /**
  * Internal dependencies
  */
-import { useStory } from '../../../../app';
+import { useStory, storyReducers } from '../../../../app';
 import useFocusCanvas from '../../../canvas/useFocusCanvas';
 
 function useLayerSelection(layer) {
   const { id: elementId } = layer;
-
-  const {
-    currentPage,
-    selectedElementIds,
-    setSelectedElementsById,
-    toggleElementInSelection,
-  } = useStory(
-    ({
-      state: { currentPage, selectedElementIds },
-      actions: { setSelectedElementsById, toggleElementInSelection },
-    }) => {
-      return {
-        currentPage,
-        selectedElementIds,
-        setSelectedElementsById,
-        toggleElementInSelection,
-      };
-    }
+  const isSelected = useStory((v) =>
+    v.state.selectedElementIds.includes(elementId)
+  );
+  const updateStateWithReducer = useStory(
+    (v) => v.actions.updateStateWithReducer
   );
 
   const focusCanvas = useFocusCanvas();
 
-  const isSelected = selectedElementIds.includes(elementId);
-  const pageElementIds = currentPage.elements.map(({ id }) => id);
-
   const handleClick = useCallback(
     (evt) => {
-      const hasSelection = selectedElementIds.length > 0;
-
       evt.preventDefault();
       evt.stopPropagation();
-      if (evt.shiftKey && hasSelection) {
-        // Shift key pressed with any element selected:
-        // select everything between this layer and the first selected layer
-        const firstId = selectedElementIds[0];
-        const firstIndex = pageElementIds.findIndex((id) => id === firstId);
-        const clickedIndex = pageElementIds.findIndex((id) => id === elementId);
-        const lowerIndex = Math.min(firstIndex, clickedIndex);
-        const higherIndex = Math.max(firstIndex, clickedIndex);
-        const elementIds = pageElementIds.slice(lowerIndex, higherIndex + 1);
-        // reverse selection if firstId isn't first anymore
-        if (firstId !== elementIds[0]) {
-          elementIds.reverse();
-        }
-        setSelectedElementsById({ elementIds });
-      } else if (evt.metaKey) {
-        // Meta pressed. Toggle this layer in the selection.
-        toggleElementInSelection({ elementId });
-      } else {
-        // No special key pressed - just selected this layer and nothing else.
-        setSelectedElementsById({ elementIds: [elementId] });
-      }
+      updateStateWithReducer({
```

Comment:
So how does this change here improve/decrease the number of re-renders?
Why is it needed? Can‘t this be avoided?
This happens at the cost of exposing reducers nonetheless.

Reply:
yea I considered prefixing this action with `dangerous` something like `dangerouslyUpdateStateWithReducer` like how react has `dangerouslySetInnerHtml`. Intent behind this is it would discourage people from using it unless they were really confident they needed it for their use case.

Basic idea tho is limiting re-renders caused by `useLayerSelection`. If we reference story state in this hook (which is ultimately just providing an action) that story state will cause a re-render in the component consuming this every time the state updates. 

Specifically here, we only want one `Layer` element to update when there's a change to an element instead of having all `Layer` elements update when there's a change to an element. If we were to pull in all story elements to reference when creating an action for each `Layer`, all `Layers` would update when a single element updated when in actuality, there was only 1 or no visual changes for the user.

As far as other approaches we could use to achieve the same benefits, we could create a new reducer that basically takes a key as it's payload and share code between the composed reducers here, however I went with the approach above cus it was the quickest way to share code and achieve the benefits mentioned here.

I imagine we'll need to compose reducers more throughout the codebase as we audit their performance, and this setup seems like a quicker win for all of those possible cases instead of writing new reducers every time we need to compose existing ones.

Reply:
@miina @barklund curious to hear your thoughts on this part too

Reply:
@littlemilkstudio Could we perhaps instead use a custom equality function here as well with `useStory`? E.g. only consider the value has changed for the currently selected elements -- `selectedElementIds.includes(elementId)`  (if I understood correctly that's the goal).

Reply:
yea I that was my first approach, just pulling the relevant data needed for this action off of `useStory` with good selectors.

The issue I ran into is that this hook is rendered in every `Layer` element, and the action this hook is supplying needs the indices of the selected elements to perform the action.

If we reference any of the selected element indices, this hook then needs to re-render every time that data updates (no matter how specific the derived data is). This would then cause every layer element to re-render whenever those indices changed (every selection change & most layer position changes).

The conclusion I came to is that all of the selected element index data needs to be derived in the scope of the reducer to achieve the needed perf constraints (renders in O(effected elements) vs O(all elements)).

I'm open to changing any implementation details, as long as we're getting the relevant state data in the closure of the reducer to prevent the original re-render issue.

Another way we could achieve this, which I'm open to if everyones opposed to this implementation, is to create a `SELECT_LAYER` action in the `useStory` reducer like: 
```
dispatch({
  type: 'SELECT_LAYER',
  payload: {
    metaKey,
    shiftKey,
    elementId
  }
})
```

And basically just make that the reducer I've created here. Thoughts?

Reply:
> If we reference any of the selected element indices, this hook then needs to re-render every time that data updates (no matter how specific the derived data is). This would then cause every layer element to re-render whenever those indices changed (every selection change & most layer position changes).

What I was thinking of is that maybe we could create an `equalityFn` for the `useStory` hook here so that it would only re-render for the selected elements (since we only need indices of the selected elements), wouldn't that work?

E.g. something like
```
const { ... } = useStory(
    ({ state, actions }) => ({
      ...
    }),
    (...) => {
      // If the element of this layer is not selected, do not re-render.
      if (hasRenderedAlready && !selectedElementIds.includes(elementId)) {
        return true;
      }
      // Some other conditions.
      ...
    };
  );
```

The example above is very simplified of course and I didn't look into it thoroughly. Perhaps you already looked into this and came to the conclusion that it wouldn't work.

Reply:
yea I guess where my thinking is diverging is that every layer element needs to know the selected elements indices for this action, not just the selected elements. This is kind of regardless of re-renders, just essential to the behavior of the action we're creating here.

If we're pulling the selected element indices through our useStory hook (however we're deriving it or altering the comparison fn), we'd need every layer to re-render if we want the functionality of the action to work.

Reply:
I guess another point I want to make is that we really shouldn't be deriving actions based off state their effecting. Kind of feels like it goes against this core principle and defeats the benefits of a reducer:
```
const [count, setCount] = useState(0);
const incrementCount = useCallback(() => setCount(count + 1), [count]);
```
vs
```
const [count, setCount] = useState(0);
const incrementCount = useCallback(() => setCount(state => state + 1), []);
```

Reply:
Alright.
@barklund Would be great to hear your thoughts here, too.

Reply:
I'm not sure why this whole thing isn't just another action inside the reducer? Why are you creating a new generic action to allow any function to be invoked and only using it once, rather than create a specific action for this specific purpose?

I hear you on the "update state based on state itself", but there are many actions already doing that - they're just handling specific situations and updating state based on the arguments given.

For this entire purpose, why not just make a new action called `toggleLayer` that takes the required arguments (probably just `isMetaPressed` and `elementId`) and move all the logic inside this action?

It would also be a lot easier to write unit tests for. This current thing here is completely untested and probably rather hard to properly test exhaustively.

Perhaps I'm missing something, that invalidates this approach?

Reply:
Quote from above that answers this question:

> As far as other approaches we could use to achieve the same benefits, we could create a new reducer that basically takes a key as it's payload and share code between the composed reducers here, however I went with the approach above cus it was the quickest way to share code and achieve the benefits mentioned here.
>
> I imagine we'll need to compose reducers more throughout the codebase as we audit their performance, and this setup seems like a quicker win for all of those possible cases instead of writing new reducers every time we need to compose existing ones.

I also think a new dedicated action for this would be cleaner.

Reply:
works for me